### ISSUE TITLE
twisted.trial.test.test_test_visitor emits deprecation warnings

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3231]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3231|
|Type|defect|
|Created|2008-05-11 22:50:51Z|
|Branch|https://github.com/twisted/twisted/tree/test-visitor-removal-3231-3|

```
exarkun@boson:~$ trial twisted.trial.test.test_test_visitor
twisted.trial.test.test_test_visitor
  TestTestVisitor
    test_visitCase ... 
/home/exarkun/Projects/Twisted/branches/system-event-ordering-3198-2/twisted/trial/unittest.py:1047: exceptions.DeprecationWarning: Test visitors deprecated in Twisted 8.0
                                                    [OK]
    test_visitEmptySuite ... 
/home/exarkun/Projects/Twisted/branches/system-event-ordering-3198-2/twisted/trial/unittest.py:1210: exceptions.DeprecationWarning: Test visitors deprecated in Twisted 8.0
                                              [OK]
    test_visitNestedSuite ...                                              [OK]
    test_visitPyunitCase ... 
/home/exarkun/Projects/Twisted/branches/system-event-ordering-3198-2/twisted/trial/unittest.py:1337: exceptions.DeprecationWarning: Test visitors deprecated in Twisted 8.0
                                              [OK]
    test_visitPyunitSuite ...                                              [OK]
    test_visitSuite ...                                                    [OK]

-------------------------------------------------------------------------------
Ran 6 tests in 0.014s

PASSED (successes=6)
exarkun@boson:~$ 
```

Attachments:

* [suppress-test_test_visitor-warnings.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1b2/1b251f64571dbb68522f1a08d3fffde0e855a306/44c4407561a245bcebae5f9f9b2d7f0f0f8ea912.patch) (784 bytes) - added by kelly on 2009-10-18 00:55:26Z - Patch to suppress the warnings.
* [remove-visitors.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1b2/1b251f64571dbb68522f1a08d3fffde0e855a306/6d7c42025754767f5316a37eb9ae0eca90df6459.patch) (12323 bytes) - added by vperic on 2012-05-28 16:46:26Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3231 3231
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_test_visitor_removal_3231_3 branches/test-visitor-removal-3231-3
branch_author__vperic__tomprince vperic, tomprince
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1210546251000000 1210546251000000
changetime__1363635139000000 1363635139000000
version__None None
owner__tom_prince tom.prince
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="kelly's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|kelly commented|
|-|-|

Patch to suppress the warnings.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml set owner to kelly|
|-|-|

Thanks for looking into this. Rather than using `suppress`, could you please use `TestCase.callDeprecated` to call the visit code. 

Also, it might actually be that the deprecation period on the visitor code has expired. I don't know how to figure this out, I'm afraid.

**Comment by twisted-trac** [comment]
|<img alt="kelly's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|kelly commented|
|-|-|

Replying to [jml](#note_3):
>
> Thanks for looking into this. Rather than using `suppress`, could you please use `TestCase.callDeprecated` to call the visit code.
>

Given that the purpose of the existing tests is to exercise the visitor code,  adding assertions about warnings is expanding their purpose somewhat.  Wouldn't it be better to add some separate tests to confirm the deprecation warnings and then suppress those warnings for the other tests?

>
> Also, it might actually be that the deprecation period on the visitor code has expired. I don't know how to figure this out, I'm afraid.
>

Looks like it was deprecated in r23027 as part of the 8.0.0 release which appears to have been released on 25-Mar-2008.  That would make this deprecation 1 year 7 months old and almost 1 major version (next release is 9.0.0?).

As far as I can tell from [the deprecation discussion](http://twistedmatrix.com/pipermail/twisted-python/2008-April/017605.html), the current policy is:

>Every deprecation must last for a minimum of 2 releases and a minimum of one year.  Of course, anyone who wants to keep maintaining deprecated APIs for longer than that may do so.

I assume 8.1 and 8.2 count as releases under this scheme?  In which case it would be eligible for removal.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

I think that --dry-run is still implemented in terms of this feature, which makes it somewhat more challenging to remove.

**Comment by twisted-trac** [comment]
|<img alt="kelly's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|kelly commented|
|-|-|

I've opened a ticket [#4063](https://github.com/twisted/twisted/issues/4063) that replaces --dry-run's use of the visitor code.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Cool.  That's taken care of, so we can proceed with this.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra commented|
|-|-|

I can't reproduce this with 11.0+?

```
bin/trial twisted.trial.test.test_test_visitor
twisted.trial.test.test_test_visitor
  TestTestVisitor
    test_visitCase ...                                                     [OK]
    test_visitEmptySuite ...                                               [OK]
    test_visitNestedSuite ...                                              [OK]
    test_visitPyunitCase ...                                               [OK]
    test_visitPyunitSuite ...                                              [OK]
    test_visitSuite ...                                                    [OK]

-------------------------------------------------------------------------------
Ran 6 tests in 0.011s

PASSED (successes=6)
```

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_9"></a>|@vperic commented|
|-|-|

Replying to [thijs](#note_8):
> I can't reproduce this with 11.0+?
> 

This is probably because you are running Python 2.7, which silences DeprecationWarnings by default. They are still there, however, so I'm attaching a patch to remove the methods (they show when running "python -3" which is mildly annoying).

This is my first time deleting some deprecated methods, hope I did everything ok.

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_11"></a>|@vperic removed owner|
|-|-|

It occurs to me now I forgot to clear the "owner" field when I submitted this for review. I'll make a branch out of it yet, but the .patch file should be enough for an initial review I hope.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set owner to @vperic|
|-|-|

[1] Some unused imports:
```
twisted/trial/test/test_pyunitcompat.py:12: 'reflect' imported but unused
twisted/trial/test/test_pyunitcompat.py:14: 'util' imported but unused
```

Please add a NEWS fragment and merge!

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_13"></a>|@vperic commented|
|-|-|

(In [34578]) Create branch test-visitor-removal-3231

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_14"></a>|@vperic commented|
|-|-|

Ok, removed the two unused imports and moved everything to a branch. Buildbot is running, [build results here](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/test-visitor-removal-3231). Will merge if everything is ok.

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_15"></a>|@vperic removed owner|
|-|-|

Hm, while fixing some docs errors, I noticed two more classes which have also been deprecated since 8.0, but haven't been tested (so didn't trigger any warnings). I've also removed them, but I guess this should go back on review. Note that one of these wasn't deprecated in the official way, but just in the docstring. Still, if no one noticed in 4 years, I guess it's not that important. :) Both of these classes were just concerned with maintaining compatibility between Python 2.3-2.5, which are no longer even supported.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve set owner to @vperic|
|-|-|

It still looks good. You should create a fragment for [#5554](https://github.com/twisted/twisted/issues/5554) and close it at the same point you're merging this.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_17"></a>|@tomprince commented|
|-|-|

(In [36848]) Branching to test-visitor-removal-3231-2

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_18"></a>|@tomprince commented|
|-|-|

I've merged this forward, but haven't looked closely to make sure everything is still sane, due to the split of `twisted.trial.unittest`.

[build results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/test-visitor-removal-3231-2)

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_19"></a>|@Julian commented|
|-|-|

Hi.

This looks good, but as per exarkun's comment, I think twisted.trial.runner.DryRunVisitor also should be removed here since [#4063](https://github.com/twisted/twisted/issues/4063) is done.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)|@tomprince set owner to @tomprince|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)|@tomprince set owner to @vperic|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_22"></a>|@tomprince commented|
|-|-|

Note: This is removing the old adaptors, which stopped being used in [#2898](https://github.com/twisted/twisted/issues/2898). There are still new adaptors that are getting used.

`DryRunVisitor` can't be removed, since it hasn't been deprecated. (I filled [#6333](https://github.com/twisted/twisted/issues/6333))

[build results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/test-visitor-removal-3231-2)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)|@tomprince removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_24"></a>|@therve set owner to @tomprince|
|-|-|

1. I have 2 tests failures in trunk (twisted.trial.test.test_loader.LoaderTest.test_loadByNamesPreservesOrder and
twisted.trial.test.test_loader.ZipLoadingTest.test_loadByNamesPreservesOrder) but I think it's easy fix, if you merge your branch forward.

2. It seems DryRunVisitor is not used anymore. I wonder if we could remove it? It's public, but I don't see what people could do with it. Otherwise open a ticket to deprecate it.

3. DryRunMixin has a suppressed warning, I think it's not needed anymore.

Please merge once fixed!

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_25"></a>|@tomprince commented|
|-|-|

(In [37465]) Branching to test-visitor-removal-3231-3.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_26"></a>|@tomprince commented|
|-|-|

1.fixed
 2. [[#6333](https://github.com/twisted/twisted/issues/6333)](https://github.com/twisted/twisted/issues/6333)
 3. That is fallout from [#4063](https://github.com/twisted/twisted/issues/4063).

As per glyphs comment on [ticket:6333#comment:5 #6333], I've also submitted [#6363](https://github.com/twisted/twisted/issues/6363) for review, so I can point to a replacement in the topfile here.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_27"></a>|@tomprince set status to `closed`|
|-|-|

(In [37594]) Merge test-visitor-removal-3231-3: Remove visitor support for trial.

Author: kelly, vperic, tom.prince
Reviewers: jml, therve
Fixes: [#3231](https://github.com/twisted/twisted/issues/3231)

twisted.trial.runner.suiteVisit and PyUnitTestCase as well
as visitor methods, all deprecated since Twisted 8.0, have
been removed.
